semantic-release misreports old builds as EGITNOPERMISSION

## Current behavior

<!-- Describe how the issue manifests. -->

We've got a fairly standard `semantic-release` setup, and started noticing that if we merge multiple PRs in quick succession or re-run an older build on our release branch that the build fails.

This isn't  a big problem as the latest commit will still succeed, and in some ways it's useful to have older builds fail (unless you want to release a fix and a feature, and still get the fix to publish under the previous minor version). It does, however, mean that our commit history looks somewhat poor and results in many issues being opened on the repo informing us that "The automated release is failing."

For example, if we have this commit sequence:

```
ce1b756 [tip]
561c792
fe34079
```

The `ce1b756` commit will not fail (and have nominal logs from `semantic-release`) no matter how many times we re-run it.

The `561c792` commit was pushed just moments before the `ce1b756` commit, and it initially failed with the `EGITNOPERMISSION` error. Attempts to re-run the build for this commit likewise fail, (have tried >3 times).

The `fe34079` commit initially succeeded (no push contention), but for diagnostic purposes we tried re-running it after the other two commits had initially failed/succeeded, and `fe34079` failed with the same `EGITNOPERMISSION`.

This would be less surprising if it were trying to do a dry-run push from `HEAD:master`, but instead the logs consistently show this:
```
[5:36:19 PM] [semantic-release] › ✖  The command "git ls-remote --heads origin master" failed with the error message remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/mixmaxhq/{redacted}.git/'
.
[5:36:19 PM] [semantic-release] › ℹ  Start step "fail" of plugin "@semantic-release/github"
```

## Expected behavior

<!-- Describe what the desired behavior would be. -->

Ideally we'd make a few things happen in response to this ticket:

- the `semantic-release` command would not fail in these circumstances, or would fail with a more informative error message
- if this is a limitation of `semantic-release`, we'd document the limitation somewhere
- if we need to fail in these circumstances, we wouldn't open issues on the repo

## Environment

- **semantic-release** version: 15.13.18
- CI environment: self-hosted Jenkins
- Plugins used: `@semantic-release/changelog`, `@semantic-release/git`
- **semantic-release** configuration:
```js
module.exports = {
  plugins: [
    '@semantic-release/commit-analyzer',
    '@semantic-release/release-notes-generator',
    '@semantic-release/github',
    '@semantic-release/npm',
    '@semantic-release/changelog',
    '@semantic-release/git',
  ],
};
```
- CI logs: see below

<details><summary>Sample of full `semantic-release` output</summary>

```
[5:36:17 PM] [semantic-release] › ℹ  Running semantic-release version 15.13.18
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/github"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/npm"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/changelog"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/git"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/release-notes-generator"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/npm"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/changelog"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/git"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/github"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/npm"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "success" from "@semantic-release/github"
[5:36:18 PM] [semantic-release] › ✔  Loaded plugin "fail" from "@semantic-release/github"
[5:36:18 PM] [semantic-release] › ✔  Run automated release from branch master
[5:36:19 PM] [semantic-release] › ✖  The command "git ls-remote --heads origin master" failed with the error message remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/mixmaxhq/{redacted}.git/'
.
[5:36:19 PM] [semantic-release] › ℹ  Start step "fail" of plugin "@semantic-release/github"
[5:36:19 PM] [semantic-release] [@semantic-release/github] › ℹ  Verify GitHub authentication
[5:36:21 PM] [semantic-release] [@semantic-release/github] › ℹ  Found existing semantic-release issue #148.
[5:36:22 PM] [semantic-release] [@semantic-release/github] › ℹ  Added comment to issue #148: https://github.com/mixmaxhq/{redacted}/issues/148#issuecomment-548487824.
[5:36:22 PM] [semantic-release] › ✔  Completed step "fail" of plugin "@semantic-release/github"
[5:36:22 PM] [semantic-release] › ✖  EGITNOPERMISSION The push permission to the Git repository is required.
semantic-release cannot push the version tag to the branch master on remote Git repository with URL https://[secure]:[secure]@github.com/mixmaxhq/{redacted}.git.

Please refer to the authentication configuration documentation (https://github.com/semantic-release/semantic-release/blob/master/docs/usage/ci-configuration.md#authentication) to configure the Git credentials on your CI environment and make sure the repositoryUrl (https://github.com/semantic-release/semantic-release/blob/master/docs/usage/configuration.md#repositoryurl) is configured with a valid Git URL (https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols).

{ SemanticReleaseError: The push permission to the Git repository is required.
    at module.exports (/{redacted}/mixmaxhq_{redacted}_master/node_modules/semantic-release/lib/get-error.js:6:10)
    at run (/{redacted}/mixmaxhq_{redacted}_master/node_modules/semantic-release/index.js:77:11)
    at process._tickCallback (internal/process/next_tick.js:68:7)
  name: 'SemanticReleaseError',
  code: 'EGITNOPERMISSION',
  details:
   '**semantic-release** cannot push the version tag to the branch `master` on remote Git repository with URL `https://[secure]:[secure]@github.com/mixmaxhq/{redacted}.git`.\n\nPlease refer to the [authentication configuration documentation](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/ci-configuration.md#authentication) to configure the Git credentials on your CI environment and make sure the [repositoryUrl](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/configuration.md#repositoryurl) is configured with a [valid Git URL](https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols).',
  semanticRelease: true }npm ERR! code ELIFECYCLE
```
</details>

--- a/index.js
+++ b/index.js
@@ -65,7 +65,7 @@ async function run(context, plugins) {
     try {
       await verifyAuth(options.repositoryUrl, options.branch, {cwd, env});
     } catch (error) {
-      if (!(await isBranchUpToDate(options.branch, {cwd, env}))) {
+      if (!(await isBranchUpToDate(options.repositoryUrl, options.branch, {cwd, env}))) {
         logger.log(
           `The local branch ${options.branch} is behind the remote one, therefore a new version won't be published.`
         );

--- a/lib/git.js
+++ b/lib/git.js
@@ -170,13 +170,14 @@ async function verifyTagName(tagName, execaOpts) {
 /**
  * Verify the local branch is up to date with the remote one.
  *
+ * @param {String} repositoryUrl The remote repository URL.
  * @param {String} branch The repository branch for which to verify status.
  * @param {Object} [execaOpts] Options to pass to `execa`.
  *
  * @return {Boolean} `true` is the HEAD of the current local branch is the same as the HEAD of the remote branch, falsy otherwise.
  */
-async function isBranchUpToDate(branch, execaOpts) {
-  const {stdout: remoteHead} = await execa('git', ['ls-remote', '--heads', 'origin', branch], execaOpts);
+async function isBranchUpToDate(repositoryUrl, branch, execaOpts) {
+  const {stdout: remoteHead} = await execa('git', ['ls-remote', '--heads', repositoryUrl, branch], execaOpts);
   try {
     return await isRefInHistory(remoteHead.match(/^(\w+)?/)[1], execaOpts);
   } catch (error) {

--- a/test/git.test.js
+++ b/test/git.test.js
@@ -214,7 +214,7 @@ test('Return "true" if repository is up to date', async t => {
   await gitCommits(['First'], {cwd});
   await gitPush(repositoryUrl, 'master', {cwd});
 
-  t.true(await isBranchUpToDate('master', {cwd}));
+  t.true(await isBranchUpToDate(repositoryUrl, 'master', {cwd}));
 });
 
 test('Return falsy if repository is not up to date', async t => {
@@ -223,13 +223,13 @@ test('Return falsy if repository is not up to date', async t => {
   await gitCommits(['Second'], {cwd});
   await gitPush(repositoryUrl, 'master', {cwd});
 
-  t.true(await isBranchUpToDate('master', {cwd}));
+  t.true(await isBranchUpToDate(repositoryUrl, 'master', {cwd}));
 
   const tmpRepo = await gitShallowClone(repositoryUrl);
   await gitCommits(['Third'], {cwd: tmpRepo});
   await gitPush('origin', 'master', {cwd: tmpRepo});
 
-  t.falsy(await isBranchUpToDate('master', {cwd}));
+  t.falsy(await isBranchUpToDate(repositoryUrl, 'master', {cwd}));
 });
 
 test('Return "true" if local repository is ahead', async t => {
@@ -238,5 +238,5 @@ test('Return "true" if local repository is ahead', async t => {
   await gitPush(repositoryUrl, 'master', {cwd});
   await gitCommits(['Second'], {cwd});
 
-  t.true(await isBranchUpToDate('master', {cwd}));
+  t.true(await isBranchUpToDate(repositoryUrl, 'master', {cwd}));
 });

